### ISSUE TITLE
Handling conflicting messages in a tipset

### ISSUE THREAD

**Issue by anorth** [filecoin-project/specs#619]
In the old world, the messages from each block were executed independently, and a block was declared invalid if any message failed to apply (didn't even get to the VM, e.g. non-existent actor, nonce too high/low, insufficient funds to pay gas, etc). When the tipset state was then evaluated, a failing message, due to conflict with a message from another block in the tipset, was simply ignored (no receipt or gas). These were rare, requiring bad luck or coordination between election-winning miners (with gas fees disincentivizing that).

With the tipset state now being the only meaningful evaluation of messages, how are to prevent blocks being stuffed with mutually invalid messages?

Some options:
- Refuse to build a tipset of mutually-incompatible blocks. Essentially drop blocks that are unlucky enough to conflict with a prior block in the same epoch.
- Approximate per-block message validity by checking each message individually against the parent state (which is not the state it will really apply to, and probably a flawed idea).
- Go back to the old world and actually execute the messages in each block individually, as well as then the whole tipset

The words currently in the [spec](https://filecoin-project.github.io/specs/#incoming-blocks-and-semantic-validation) describe only the old world.

See related #600

**Comment by whyrusleeping** [comment]
The tipset state changes don't really change this much. You can still force individual blocks to not have internally conflicting messages (no duplicate/bad nonces, balance checks). For miners to actually have significant amounts of messages conflicting between blocks in a tipset, they would have to either have severely out of sync message pools, or be doing this intentionally. In either case, I think that the market will work itself out. Doing that is really just preventing yourself from getting gas fees. This is the same class of 'attack' that filling your own blocks with spam transactions from yourself to yourself with zero gas cost would be. Sure you could do it, but you're missing out.

**Comment by anorth** [comment]
> You can still force individual blocks to not have internally conflicting messages

Right, that's the "old world" solution. Note that effective balance checks require actually executing all the messages, otherwise the first message can consume all the balance and invalidate all the others, which might individually be ok when evaluated directly against the parent state.

So, will the spec require that individual blocks don't have conflicting messages?

**Comment by anorth** [comment]
cc @acruikshank @ZenGround0

**Comment by whyrusleeping** [comment]
>  Note that effective balance checks require actually executing all the messages,

No, thats not true. You just check the message value and gas requirements. Check how lotus does this here: https://github.com/filecoin-project/lotus/blob/master/miner/miner.go#L332

**Comment by acruikshank** [comment]
@whyrusleeping This code doesn't account for crerdits to an account that later sends a message whose gas limit would have otherwise exceed funds. This seems like a fine strategy for selecting messages for block generation, but we probably need an invariant that quick validation fails with insufficient gas iff they would have failed for the same reason during processing. 

If we were to validate this way, we would need to make explicit in the spec that a message is invalid if its funds in the parent state balance minus all its debits (but not credits) is below the gas limit. This will force us to run the quick validation whether or not we process messages.

**Comment by anorth** [comment]
> check the message value and gas requirements

Thanks, my mistake.

Request for the spec remains that, if there are message sequence checks that will render a block invalid, to define all those checks and confirm they do not require evaluating messages.

FTR, here are the list of reasons go-filecoin defines that cause a message to be invalid:
errFromAccountNotFound
errNonceTooHigh
errGasTooHighForCurrentBlock
errInsufficientGas
errSelfSend
errNonceTooLow
errNonAccountActor
errNegativeValue
errCannotTransferNegativeValue
errGasAboveBlockLimit

**Comment by jbenet** [comment]
- If a message is in a block, it doesn't matter that it may conflict, the VM interpreter can take it and should either execute correctly, or (correctly) fail to execute. it is not _invalid_ and has no bearing in block validity. it just fails to execute when it hits the VM.
- Messages with a bad nonce or insufficient balance to cover their gas (gas_limit * gas_price) are not _invalid_, they just fail when applying.
  - VM interpreter `ApplyMessage` exits early, not changing any state, not consuming any gas
- Messages which _run out of gas_ or which _fail with an error_ are not invalid, and are invoked, but fail with an error. the state is unmodified except for balance (gas is indeed spent) and nonce (nonce is incremented -- the message did apply)

see https://filecoin-project.github.io/specs/#vm-interpreter-implementation

> doesn't account for crerdits to an account that later sends a message whose gas limit would have otherwise exceed funds.

if msg1 affects state to enable msg2 to run that's perfectly OK. it is not invalid, it is not an error. Just as if msg3 affected state to disable msg4 from running. msg4 would fail to apply and end in an error (after applying, still not invalid)

to be explicit:
- tipset = [block1, block2]
- msg1 in block1 from account1 sends money to account2 (from 0 -> 100)
  - before msg1 lands, account2 does not have any money for gas
- msg2 in block2 executes some function, spending gas.
  - msg2 applies after msg1 and therefore applies fine because now there's money.

> This seems like a fine strategy for selecting messages for block generation, but we probably need an invariant that quick validation fails with insufficient gas iff they would have failed for the same reason during processing.
>
> If we were to validate this way, we would need to make explicit in the spec that a message is invalid if its funds in the parent state balance minus all its debits (but not credits) is below the gas limit. This will force us to run the quick validation whether or not we process messages.

No we really should not be running checks like that and removing messages before they hit the VM. Only the VM processing the StateTree is the arbiter of whether a message applies or not. Removing it early is error prone and likely to cause many more problems than it theoretically helps with. It will force a fake pre-computation of state to move out of the VM, and duplicate logic, liable to be wrong. 

I'll make sure there's a test in a test chain that correctly checks for this (msg1/msg2 case above)

**Comment by acruikshank** [comment]
@jbenet super clear. Thanks.

**Comment by anorth** [comment]
> it is not invalid and has no bearing in block validity

This is clear, but I also want to be explicit that this contradicts @whyrusleeping:
> You can still force individual blocks to not have internally conflicting messages 

(You *can*, but we're not going to).

Nothing will prevent a miner producing a block with messages with nonsensical nonces, nonexistent actors etc, which consume no gas and produce no receipt. (Ack that it's not first-order rational to do so).

The spec (carried over from the old spec) is conflicting about this. Regarding block semantic validity, it [says](https://filecoin-project.github.io/specs/#incoming-blocks-and-semantic-validation):

> - must only have valid state transitions:
>     - all messages in the block must be valid
>     - the execution of each message, in the order they are in the block, must produce a receipt matching the corresponding one in the receipt set of the block.

**Comment by whyrusleeping** [comment]
I think its absolutely worth filtering out messages that are not consistent within a single block. If a block contains two messages with the same sender and same nonce, that block should be invalid. We should not reward miners for adding that extra (easily filterable) garbage to the chain.
